plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,1059 @@
+  username = "admin"
+
+  ## The user's password to authenticate with
+  password = "Passw0rd"
+
+  ## Services to be enabled
+  #enabled_services = ["stacks","services", "projects", "hypervisors", "flavors", "servers", "volumes", "storage" , "subnets", "ports", "networks", "aggregates", "nova_services", "agents"]
+  enabled_services = ["services", "projects", "hypervisors", "flavors", "networks", "volumes"]
+
+  #Dependencies
+  # | Service | Depends on |
+  # | servers | projects, hypervisors, flavors |
+  # | volumes | projects |
+
+  ## Collect Server Diagnostics
+  server_diagnotics = false
+
+  InsecureSkipVerify = false
+`
+
+// SampleConfig return a sample configuration file for auto-generation and
+// implements the Input interface.
+func (o *OpenStack) SampleConfig() string {
+	return sampleConfig
+}
+
+// Gather gathers resources from the OpenStack API and accumulates metrics.  This
+// implements the Input interface.
+func (o *OpenStack) Gather(acc telegraf.Accumulator) error {
+	// Perform any required set up
+	if err := o.initialize(); err != nil {
+		return err
+	}
+
+	// Gather resources.  Note service harvesting must come first as the other
+	// gatherers are dependant on this information.
+	gatherers := map[string]func() error{
+		"projects":      o.gatherProjects,
+		"hypervisors":   o.gatherHypervisors,
+		"flavors":       o.gatherFlavors,
+		"servers":       o.gatherServers,
+		"volumes":       o.gatherVolumes,
+		"storage pools": o.gatherStoragePools,
+		"subnets":       o.gatherSubnets,
+		"ports":         o.gatherPorts,
+		"networks":      o.gatherNetworks,
+		"aggregates":    o.gatherAggregates,
+		"nova_services": o.gatherNovaServices,
+		"agents":        o.gatherAgents,
+		"stacks":        o.gatherStacks,
+	}
+	for resources, gatherer := range gatherers {
+		for _, i := range o.EnabledServices {
+			if resources == i {
+				if err := gatherer(); err != nil {
+					log.Println("W!", plugin, "failed to get", resources, err)
+				}
+			}
+		}
+	}
+
+	// Accumulate statistics
+	accumulators := map[string]func(telegraf.Accumulator){
+		"services":      o.accumulateServices,
+		"projects":      o.accumulateIdentity,
+		"hypervisors":   o.accumulateHypervisors,
+		"flavors":       o.accumulateFlavors,
+		"servers":       o.accumulateServers,
+		"volumes":       o.accumulateVolumes,
+		"storage pools": o.accumulateStoragePools,
+		"subnets":       o.accumulateSubnets,
+		"ports":         o.accumulatePorts,
+		"networks":      o.accumulateNetworks,
+		"aggregates":    o.accumulateAggregates,
+		"nova_services": o.accumulateNovaServices,
+		"agents":        o.accumulateAgents,
+		"stacks":        o.accumulateStacks,
+	}
+	for resources, accumulator := range accumulators {
+		for _, i := range o.EnabledServices {
+			if resources == i {
+				accumulator(acc)
+			}
+		}
+	}
+	return nil
+}
+
+// initialize performs any necessary initialization functions
+func (o *OpenStack) initialize() error {
```

Comment:
Just rename this to
```suggestion
func (o *OpenStack) Init() error {
```
and telegraf will automatically call this function after all user-config options are set.